docker/bookkeeper/entrypoint.sh

```diff
@@ -48,31 +56,151 @@ export BK_tlsTrustStoreType=JKS
+   # an error code `255`. Since, the `bookkeeper/bin/bookeeper` shell script sources the `common.sh` file, and it sets
+   # "exit on error" as true, we encounter this error when it is invoked. To avoid that error, here we modify the
+   # `common.sh` file to execute `/sbin/sysctl -n net.ipv6.bindv6only` only if file `/proc/sys/net/ipv6/bindv6only`
+   # is available.
+
+   # Replace the 22nd line with "if [ -f /sbin/sysctl ] && [ -f /proc/sys/net/ipv6/bindv6only ]; then".
+   sed -i "22s|.*|if [ -f /sbin/sysctl ] \&\& [ -f /proc/sys/net/ipv6/bindv6only ]; then|" ${BK_HOME}/bin/common.sh
+}
+
+initialize_cluster() {
+    set +e
+
+    zk-shell --run-once "ls ${BK_zkLedgersRootPath}/available/readonly" ${BK_zkServers}
+    if [ $? -eq 0 ]; then
+        echo "Cluster metadata already exists"
+    else
+        # Create an ephemeral zk node `bkInitLock` for use as a lock.
+        lock=`zk-shell --run-once "create ${BK_CLUSTER_ROOT_PATH}/bkInitLock '' true false false" ${BK_zkServers}`
+        if [ -z "$lock" ]; then
+            echo "Bookkeeper znodes do not exist in Zookeeper. Initializing a new Bookeekeper cluster."
+
+            # Note that this `bookkeeper` shell script sets "exit on error" (`set -e`) and sources from
+            # ${SCRIPTS_DIR}/common.sh. Make sure to invoke `fix_bk_ipv6_check()` before the control reaches here.
+            /opt/bookkeeper/bin/bookkeeper shell initnewcluster
+            if [ $? -eq 0 ]; then
+                echo "initnewcluster operation succeeded"
+            else
+                echo "initnewcluster operation failed. Please check the reason."
+                echo "Exit status of initnewcluster"
+                echo $?
+                exit
+            fi
+        else
+            echo "Others may be initializing the cluster at the same time."
+            tenSeconds=1
+            while [ ${tenSeconds} -lt 100 ]
+            do
+                sleep 10
+                zk-shell --run-once "ls ${BK_zkLedgersRootPath}/available/readonly" ${BK_zkServers}
+                if [ $? -eq 0 ]; then
+                    echo "Waited $tenSeconds * 10 seconds. Successfully listed ''${BK_zkLedgersRootPath}/available/readonly'"
+                    break
+                else
+                    echo "Waited $tenSeconds * 10 seconds. Continue waiting."
+                    (( tenSeconds++ ))
+                    continue
+                fi
+            done
+
+            if [ ${tenSeconds} -eq 100 ]; then
+                echo "Waited 100 seconds for bookkeeper cluster to initialize, but to no avail. Something is wrong, please check."
+                exit
+            fi
+        fi
+    fi
+    set -e
+}
+
+format_bookie_data_and_metadata() {
+    if [ `find $BK_journalDirectory $BK_ledgerDirectories $BK_indexDirectories -type f 2> /dev/null | wc -l` -gt 0 ]; then
+      # The container already contains data in BK directories. Examples of when this can happen include:
```

Comment:
Say there is some kind of partial loss and some of the directories contain data but others lost it. What is the expected behavior in this situation?

Reply:
In a scenario where a subset of these directories has lost data, based on the code, the `else` block is not executed. So, 
* The current bookie's server contents are not deleted
* The current bookie's cookie is not deleted in Zookeeper

The expected behavior would be that the bookie might fail to start-up unless the directory that was lost is in a recoverable state. 

It should also be noted that this is a preexisting code that was extracted into a function to make it easier to read.

Reply:
> The expected behavior would be that the bookie might fail to start-up unless the directory that was lost is in a recoverable state.

That's right, see the comment in the BK code:

https://github.com/apache/bookkeeper/blob/master/bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Cookie.java#L59

This is not new code, so let me create an issue to investigate this.